lib/spack/spack/filesystem_view.py

```diff
@@ -398,23 +398,32 @@ def remove_specs(self, *specs, **kwargs):
                      "The following packages will be unusable: %s"
                      % ", ".join((s.name for s in dependents)))
 
-        extensions = set(filter(lambda s: s.package.is_extension,
-                         to_deactivate))
-        standalones = to_deactivate - extensions
-
-        # Please note that a traversal of the DAG in post-order and then
-        # forcibly removing each package should remove the need to specify
-        # with_dependents for deactivating extensions/allow removal without
-        # additional checks (force=True). If removal performance becomes
-        # unbearable for whatever reason, this should be the first point of
-        # attack.
-        #
-        # see: https://github.com/spack/spack/pull/3227#discussion_r117147475
-        remove_extension = ft.partial(self.remove_extension,
-                                      with_dependents=with_dependents)
-
-        set(map(remove_extension, extensions))
-        set(map(self.remove_standalone, standalones))
+        # depmap maps each package to its list of dependencies. This is used to
+        # sort the packages into dependency post-order.
+        depmap = dict()
+        for spec in to_deactivate:
+            depmap[spec] = set(d for d in spec.traverse(root=False)
+                               if d in to_deactivate)
+
+        # to_deactivate_sorted is the packages to deactivate in post-order.
+        to_deactivate_sorted = []
+        while depmap:
+            for spec in [s for s, d in depmap.items() if not d]:
+                to_deactivate_sorted.append(spec)
+                for s in depmap.keys():
+                    depmap[s].discard(spec)
+                depmap.pop(spec)
+        to_deactivate_sorted.reverse()
```

Comment:
In terms of the terminology of `Spack.Spec.traverse` if we have a DAG like

```
x->y->z
```

where x depends on y and y depends on z, post-order means [z, y, x]. I'm not sure if you meant something different, but after this reversal `to_deactivate_sorted` is a preordered traversal. That is correct but I think the comments (e.g. lines 401 and 408) should be updated. I think the following comment would be clearest:

> determine the order that packages should be removed from the view; dependents come before their dependencies.

at line 401, and for clarity you could relocate the declaration of `to_deactivate_sorted` there, like:

```
# determine the order that packages should be removed from the view; dependents come before their dependencies.
to_deactivate_sorted = list()

depmap = dict()
...
to_deactivate_sorted.reverse()
```

Does that seem reasonable?

Reply:
Yep sounds good. That is what I think I was trying to convey - I always find it confusing to get pre/post child/parent dependents/dependencies the right way around!